prog/dftb+/lib_extlibs/negf_int.F90

```diff
@@ -0,0 +1,2117 @@
+! *
+! *  call negf_current(HH,SS,spin,kpoint,wght,tunn,ledos,currents)
+! *
+! *  call negf_destroy()
+! *  ----------------------------------------------------------------------
+! *
+! *  ---------------------------------------------------------------------- 
+! * Known Bugs & Problems:
+! * 
+! *
+! *
+! * ---------------------------------------------------------------------- 
+module negf_int
+  
+  use accuracy
+  use libnegf_vars  !, only : TNEGFstructure, TGDFTBTunDos, &
+                    !       & TNEGFGreenDensInfo, TTransPar, Telph
+  use libnegf
+  use mat_conv
+  use sparse2dense
+  use densedescr
+  use commonTypes, only : TOrbitals
+  use libmpifx_module
+
+  use FormatOut                                                          
+  
+  implicit none
+  private
+
+  Type(Tnegf), target, public :: negf
+
+  public :: negf_init      ! general library initializations
+  public :: negf_init_str  ! passing structure parameters
+  public :: negf_init_dephasing
+  public :: negf_init_elph                                               
+  public :: negf_destroy
+  
+  ! wrapped functions passing dftb matrices. Needed for parallel
+  public :: calcdensity_green
+  public :: calcEdensity_green
+  public :: calcPDOS_green
+  public :: calc_current
+  public :: local_currents
+  
+  ! interface csr matrices. The pattering must be predefined 
+  ! using negf_init_csr 
+  public :: negf_init_csr
+  type(z_CSR) :: csrHam, csrOver
+
+  ! non wrapped direct calls  
+  private :: negf_density, negf_current, negf_ldos
+
+  contains
+
+  !------------------------------------------------------------------------------
+  ! Init gDFTB environment and variables
+  !
+  ! Note: mpicomm should be the global commworld here
+  !------------------------------------------------------------------------------
+  subroutine negf_init(transpar, greendens, tundos, mpicomm,&
+        & tempElec, initinfo)
+    
+    Type(TTranspar), intent(IN) :: transpar
+    Type(TNEGFGreenDensInfo), intent(IN) :: greendens
+    Type(TNEGFTunDos), intent(IN) :: tundos
+    Type(mpifx_comm), intent(in) :: mpicomm
+    real(dp), intent(in) :: tempElec 
+    logical, intent(OUT) :: initinfo
+    
+    ! local variables
+    real(dp), dimension(:), allocatable :: pot, eFermi
+    integer :: i,error, l, ncont, nc_vec(1), j, nldos
+    integer, dimension(:), allocatable :: sizes  
+    ! string needed to hold processor name
+    character(:), allocatable :: hostname
+    type(lnParams) :: params
+   
+    error = 0 
+    initinfo = .true.       
+
+    call negf_mpi_init(mpicomm)
+     
+    if (transpar%defined) then
+      ncont = transpar%ncont 
+    else
+      ncont = 0
+    endif
+    
+    ! ------------------------------------------------------------------------------
+    !! Set defaults and fill up the parameter structure with them
+    call init_negf(negf)
+    call init_contacts(negf, ncont)
+    call set_scratch(negf, ".")
+    if (id0) call create_scratch(negf)    
```

Comment:
id0 is coming back from libNEGF, probably should be using internal DFTB+ variables (there is a logical in the environment module for this).